### PR TITLE
Resolve filepaths before scanning for changes

### PR DESCRIPTION
**Pull request by cvgw**
**Description**

Implement the new filepath resolver described in https://github.com/GoogleContainerTools/kaniko/blob/d31d2bc74c884ff5d7932b1f6e7209c6fc63cc5e/docs/design_proposals/filesystem-resolution-proposal-01.md

One change to note is that ancestor directories will no longer be added regardless of whether they are changed. The ancestor directory logic now occurs before the change checking, so ancestor directories will only be added if they have changed.

**Submitter Checklist** 

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

### CODE REVIEW

File: pkg/snapshot/snapshot_test.go [hunk 1]
```diff
@@ -63,19 +63,24 @@ func TestSnapshotFSFileChange(t *testing.T) {
...
 	}
-	for _, dir := range util.ParentDirectoriesWithoutLeadingSlash(fooPath) {
-		snapshotFiles[dir] = ""
-	}
-	for _, dir := range util.ParentDirectoriesWithoutLeadingSlash(batPath) {
-		snapshotFiles[dir] = ""
-	}
-	numFiles := 0
+
+	// Their parents didn't change so they shouldn't be included
```
Comment (samos123) [inline pkg/snapshot/snapshot_test.go]: I would prefer to remove the code, we can always recover it from git.

File: pkg/snapshot/snapshot_test.go [hunk 2]
```diff
@@ -155,17 +160,20 @@ func TestSnapshotFSChangePermissions(t *testing.T) {
 	snapshotFiles := map[string]string{
 		batPathWithoutLeadingSlash: "baz2",
 	}
-	for _, dir := range util.ParentDirectoriesWithoutLeadingSlash(batPath) {
-		snapshotFiles[dir] = ""
-	}
-	numFiles := 0
+
+	// The parents haven't changed so they shouldn't be in the tar
+	//for _, dir := range util.ParentDirectoriesWithoutLeadingSlash(batPath) {
```
Comment (samos123) [inline pkg/snapshot/snapshot_test.go]: same here, remove the code instead of commenting it out

File: pkg/filesystem/resolve.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// filesWithParentDirs returns every ancestor path for each provided file path.
+// I.E. /foo/bar/baz/boom.txt => [/, /foo, /foo/bar, /foo/bar/baz, /foo/bar/baz/boom.txt]
+func filesWithParentDirs(files []string) []string {
+	filesSet := map[string]bool{}
+
+	for _, file := range files {
+		file = filepath.Clean(file)
+		filesSet[file] = true
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: you can skip traversing if fileSet[file] = true
```suggestion
		if _, ok := fileSet[file] ; !ok {
		   filesSet[file] = true
		   for _, dir := range util.ParentDirectories(file) {
			dir = filepath.Clean(dir)
			filesSet[dir] = true
		   }
		 }
```

File: pkg/filesystem/resolve.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+		for _, dir := range util.ParentDirectories(file) {
+			dir = filepath.Clean(dir)
+			filesSet[dir] = true
+		}
+	}
+
+	newFiles := []string{}
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: We can optimize this by allocation memory of expected size
```
newFiles := make([]string, 0, len(filesSet))
```

File: pkg/snapshot/snapshot.go [hunk 5]
```diff
@@ -149,19 +155,42 @@ func (s *Snapshotter) scanFullFilesystem() ([]string, []string, error) {
 	)
 	timing.DefaultRun.Stop(timer)
 
+	filesToResolve := make([]string, 0, len(memFs))
+	for file := range memFs {
+		if strings.HasPrefix(file, "/tmp/dir") {
```
Comment (tejal29) [inline pkg/snapshot/snapshot.go]: is this intentional change?

File: pkg/filesystem/resolve.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	if !filepath.IsAbs(path) {
+		return "", errors.New("dest path must be abs")
+	}
+	last := ""
+	newPath := path
+loop:
+	for newPath != "/" {
+		fi, err := os.Lstat(newPath)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to lstat")
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: does the err contain the file name? if not, please add the file name

File: pkg/filesystem/resolve.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		}
+
+		switch mode := fi.Mode(); {
+		case mode&os.ModeSymlink != 0:
+			last = filepath.Base(newPath)
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: you can use `util.IsSymlink` here instead.

File: pkg/filesystem/resolve.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+			newPath = filepath.Dir(newPath)
+		default:
+			target, err := filepath.EvalSymlinks(newPath)
+			if err != nil {
+				return "", err
+			}
+
+			if target != newPath {
+				last = filepath.Base(newPath)
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: shd this be target?

File: pkg/filesystem/resolve.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+		fi, err := os.Lstat(newPath)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to lstat")
+		}
+
+		switch mode := fi.Mode(); {
+		case mode&os.ModeSymlink != 0:
+			last = filepath.Base(newPath)
+			newPath = filepath.Dir(newPath)
+		default:
```
Comment (tejal29) [inline pkg/filesystem/resolve.go]: I am a little bit lost here and probably we need to add tests here. 
say a/b/foo.txt is a link pointing to c/d/bar.txt.

In this case, we need to follow the symlink for `a/b/foo.txt` and in your case the `default` block follows the symlink for `a/b/`

Is my analysis correct?

File: pkg/filesystem/resolve.go (hunk 6 above)
Comment (cvgw) [reply pkg/filesystem/resolve.go]: yes, lstat errors always contain the file name

File: pkg/filesystem/resolve.go (hunk 8 above)
Comment (cvgw) [reply pkg/filesystem/resolve.go]: I don't believe so. We want to return the link, not the target.

File: pkg/snapshot/snapshot.go (hunk 5 above)
Comment (cvgw) [reply pkg/snapshot/snapshot.go]: Nope, left over cruft from debugging! Thanks for catching

File: pkg/snapshot/snapshot_test.go (hunk 1 above)
Comment (cvgw) [reply pkg/snapshot/snapshot_test.go]: Yes, plan is to remove after testing it finished

File: pkg/filesystem/resolve.go (hunk 3 above)
Comment (cvgw) [reply pkg/filesystem/resolve.go]: The function `filesWithParentDirs` was copy/pasted from the util pkg. If you think it won't clutter the change I'm happy to take this opportunity to make these updates.

File: pkg/filesystem/resolve.go (hunk 9 above)
Comment (cvgw) [reply pkg/filesystem/resolve.go]: In your example the default block would not be hit as `a/b/foo.txt` is a symlink (as opposed to one of it's ancestors) so the return of this function would be `a/b/foo.txt`

If, for example `a/b/` was a link to `c/d/` then `a/b/bar.txt` would return `a/b/`

Adding tests

Comment (tejal29) [reply pkg/filesystem/resolve.go]: ah got it! Thanks. the `if else` makes it more clear.

**Comment by cvgw** [comment]
related issues #1039 #1038 #1024

**Pull request merged**